### PR TITLE
Add ads api support [WIP]

### PR DESCRIPTION
**Pull request by thomas-holmes**
*Accounts*
 - [X] GET     accounts/:account_id
 - [X] GET     accounts

*Campaigns*
 - [X] GET     accounts/:account_id/campaigns
 - [X] GET     accounts/:account_id/campaigns/:campaign_id
 - [X] POST    accounts/:account_id/campaigns
 - [X] PUT     accounts/:account_id/campaigns/:campaign_id
 - [X] DELETE  accounts/:account_id/campaigns/:campaign_id

*Line Items*
 - [X] GET     accounts/:account_id/line_items/:line_item_id
 - [X] GET     accounts/:account_id/line_items
 - [X] POST    accounts/:account_id/line_items
 - [X] PUT     accounts/:account_id/line_items/:line_item_id
 - [X] DELETE  accounts/:account_id/line_items/:line_item_id

*Promoted Accounts*
 - [X] GET     accounts/:account_id/promoted_accounts
 - [X] POST    accounts/:account_id/promoted_accounts

*Promotable Users*
 - [X] GET     accounts/:account_id/promotable_users

*Promoted Tweets*
 - [X] GET     accounts/:account_id/promoted_tweets
 - [X] POST    accounts/:account_id/promoted_tweets
 - [X] DELETE  accounts/:account_id/promoted_tweets/:id

*Funding Instruments*
 - [X] GET     accounts/:account_id/funding_instruments
 - [X] GET     accounts/:account_id/funding_instruments/:id

*Targeting Suggestions*
 - [x] GET     accounts/:account_id/targeting_suggestions

*Reach Estimate*
 - [ ] GET     accounts/:account_id/reach_estimate

*Targeting Criteria*
 - [x] GET     accounts/:account_id/targeting_criteria
 - [x] GET     accounts/:account_id/targeting_criteria/:id
 - [x] POST    accounts/:account_id/targeting_criteria
 - [x] PUT     accounts/:account_id/targeting_criteria
 - [x] DELETE  accounts/:account_id/targeting_criteria/:id
 - [x] GET     targeting_criteria/app_store_categories
 - [x] GET     targeting_criteria/behavior_taxonomies
 - [x] GET     targeting_criteria/behaviors
 - [x] GET     targeting_criteria/devices
 - [x] GET     targeting_criteria/interests
 - [x] GET     targeting_criteria/languages
 - [x] GET     targeting_criteria/locations
 - [x] GET     targeting_criteria/network_operators
 - [x] GET     targeting_criteria/platform_versions
 - [x] GET     targeting_criteria/platforms
 - [x] GET     targeting_criteria/tv_channels
 - [x] GET     targeting_criteria/tv_genres
 - [x] GET     targeting_criteria/tv_markets
 - [x] GET     targeting_criteria/tv_shows

*Tailored Audiences*
 - [X] GET     accounts/:account_id/tailored_audiences
 - [X] GET     accounts/:account_id/tailored_audiences/:id
 - [X] POST    accounts/:account_id/tailored_audiences
 - [X] DELETE  accounts/:account_id/tailored_audiences/:id
 - [X] GET     accounts/:account_id/tailored_audience_changes
 - [X] GET     accounts/:account_id/tailored_audience_changes/:id
 - [X] POST    accounts/:account_id/tailored_audience_changes
 - [x] PUT     accounts/:account_id/tailored_audiences/global_opt_out

*Tweet!*
 - [x] POST    accounts/:account_id/tweet  Beta

*Mobile App Cards*
 - [x] GET     accounts/:account_id/cards/app_download
 - [x] GET     accounts/:account_id/cards/app_download/:card_id
 - [x] POST    accounts/:account_id/cards/app_download
 - [x] PUT     accounts/:account_id/cards/app_download/:card_id
 - [x] DELETE  accounts/:account_id/cards/app_download/:card_id

*Mobile App (Image) Cards*
 - [x] GET     accounts/:account_id/cards/image_app_download
 - [x] GET     accounts/:account_id/cards/image_app_download/:card_id
 - [x] POST    accounts/:account_id/cards/image_app_download
 - [x] PUT     accounts/:account_id/cards/image_app_download/:card_id
 - [x] DELETE  accounts/:account_id/cards/image_app_download/:card_id

*Lead Gen Cards*
 - [x] GET     accounts/:account_id/cards/lead_gen
 - [x] GET     accounts/:account_id/cards/lead_gen/:card_id
 - [x] POST    accounts/:account_id/cards/lead_gen
 - [x] PUT     accounts/:account_id/cards/lead_gen/:card_id
 - [x] DELETE  accounts/:account_id/cards/lead_gen/:card_id

*Website Cards*
 - [x] GET     accounts/:account_id/cards/website
 - [x] GET     accounts/:account_id/cards/website/:card_id
 - [x] POST    accounts/:account_id/cards/website
 - [x] PUT     accounts/:account_id/cards/website/:card_id
 - [x] DELETE  accounts/:account_id/cards/website/:card_id

*App Events*
 - [ ] GET     accounts/:account_id/app_event_provider_configurations
 - [ ] GET     accounts/:account_id/app_event_provider_configurations/:id
 - [ ] GET     accounts/:account_id/app_event_tags
 - [ ] GET     accounts/:account_id/app_event_tags/:id

*Web Events*
 - [ ] GET     accounts/:account_id/web_event_tags
 - [ ] GET     accounts/:account_id/web_event_tags/:web_event_tag_id
 - [ ] POST    accounts/:account_id/web_event_tags
 - [ ] PUT     accounts/:account_id/web_event_tags/:web_event_tag_id
 - [ ] DELETE  accounts/:account_id/web_event_tags/:web_event_tag_id

*Scoped Timeline*
 - [x] GET     accounts/:account_id/scoped_timeline

*Bidding Rules*
 - [ ] GET     bidding_rules

*Card Stats*
 - [ ] GET     stats/accounts/:account_id/cards/lead_gen/:card_id
 - [ ] GET     stats/accounts/:account_id/cards/website/:card_id
 - [ ] GET     stats/accounts/:account_id/cards/app_download/:card_id
 - [ ] GET     stats/accounts/:account_id/cards/image_app_download/:card_id

*Campaign Stats*
 - [ ] GET     stats/accounts/:account_id/campaigns/:id
 - [ ] GET     stats/accounts/:account_id/campaigns

*Funding Instrument Stats*
 - [ ] GET     stats/accounts/:account_id/funding_instruments/:id
 - [ ] GET     stats/accounts/:account_id/funding_instruments

*Line Item Stats*
 - [ ] GET     stats/accounts/:account_id/line_items/:id
 - [ ] GET     stats/accounts/:account_id/line_items

*Promoted Account Stats*
 - [ ] GET     stats/accounts/:account_id/promoted_accounts/:id
 - [ ] GET     stats/accounts/:account_id/promoted_accounts

*Promoted Tweets Stats*
 - [ ] GET     stats/accounts/:account_id/promoted_tweets/:id
 - [ ] GET     stats/accounts/:account_id/promoted_tweets

*Account Stats*
 - [ ] GET     stats/accounts/:account_id

### CODE REVIEW

**Comment by thomas-holmes** [comment]
There's still a lot of work to do here. I think some refactoring is going to be required but I took the path of least resistance initially so I could get some feedback.

I've tried to maintain the existing structure. I'm placing the Ads API models in `lib/twitter` and the request oriented things in `lib/twitter/ads`.

I've had to duplicate some code as well (rather than refactoring at this point) to support some structural changes in the way the API works.

1. The Ads API appears to respond to every request wrapped in a "data" key and also returns the request params as well. As a stop-gap I've duplicated `Twitter::Rest::Utils` to `Twitter::Ads::Utils` for essentially a very small change.

2. I'm not able to get some portion of the sandbox working at the moment so I am unable to test (or implement) cursoring. The existing method should probably work but the cursor values are not integers so the logic would need to change a little bit (defaulting to -1 probably is not correct behavior for the Ads API).

3. A lot of the endpoints are nested and require multiple ids to function  e.g. `GET /0/accounts/:account_id/campaigns/:campaign_id`. This makes the interface feel a bit awkward. For now I've chosen to just implement the client as `@client.campaign(account_id, campaign_id)`.

4. I'm currently reusing the `Twitter::Identity` class. The in-code documentation specifies `@return [Integer]`. The Ads API uses alphanumeric identifiers. For now I'm just using the existing `Twitter::Identity` class but the docs are lying. I figure I can just update the documentation to indicate `@return [Integer,String]` but I thought you might also want to keep them separate so a consumer can read the docs and understand the id characteristics.

I've also added an ads.md file to the root of the project to keep track of my progress through implementing the ads endpoints. It's been sorted in a slightly less insane format than the documentation website.

File: spec/twitter/ads/line_items_spec.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    let(:args) do
+      {
+        bid_amount_local_micro: 100_000,
+        placement_type: 'PROMOTED_TWEETS_FOR_SEARCH',
+        objective: 'TWEET_ENGAGEMENTS',
+        paused: true,
+      }
+    end
+    let(:campaign_id) { '8lp0' }
+    let(:expected) {args.map {|k,v| [k, v.to_s]}.to_h.merge(campaign_id: '8lp0')}
```
Comment (thomas-holmes) [inline spec/twitter/ads/line_items_spec.rb]: Whoops, forgot this wasn't available in older Rubies.

**Comment by thomas-holmes** [comment]
The oauth tokens are cross functional (interop between ads and original REST api)  so I'm thinking about integrating the two clients together again. This will also make things nicer for the user (when dealing with promotting accounts/tweets) because the ads structures refer to traditional twitter resources (tweets, users).

Also I think I'm going to have to add some TON support, at least enough to work with Tailored Audiences.

File: ads.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+| | GET    | stats/accounts/:account_id/line_items/:id |
+| | GET    | stats/accounts/:account_id/line_items |
+||| *Promoted Account Stats* |
+| | GET    | stats/accounts/:account_id/promoted_accounts/:id |
+| | GET    | stats/accounts/:account_id/promoted_accounts |
+||| *Promoted Tweets Stats* |
+| | GET    | stats/accounts/:account_id/promoted_tweets/:id |
+| | GET    | stats/accounts/:account_id/promoted_tweets |
+||| *Account Stats* |
+| | GET    | stats/accounts/:account_id |
```
Comment (sferik) [inline ads.md]: Does it make sense to move this file out of source control and into [a GitHub Task List](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments) on this pull request? I think that will make it easier to track and update progress.

Comment (thomas-holmes) [inline ads.md]: Probably. I just had it there for intermittent work because it didn't used to live inside of a PR. I'll convert it later.

**Comment by thomas-holmes** [comment]
Ok I moved it up into the top. That's a heck of an opening comment :)

**Comment by ianks** [comment]
@thomas-holmes do you want help with this?

**Comment by thomas-holmes** [comment]
So I just added all of the stats endpoints though I have a few pending issues.

1. The requests don't seem to work even though I can make requests with Twurl. I haven't gotten to the bottom of this yet.
2. The stats endpoints are rate limited by cost on a rolling schedule. This information, as far as I know, is only available in response headers. I'm not sure where to place it. Currently I'm just shoving stats into a `Twitter::Stats` object, but I'm thinking maybe I should make a stats container that saves the header values and them embed the actual stat object/hashes inside of it. Thoughts?

**Comment by thomas-holmes** [comment]
So I've somewhat tracked down the issue: It seems that [httprb/http](https://github.com/httprb/http) versions >= 8.0 change something in the way query parameters are encoded which causes a mismatch during oauth signing and validation.

The specific case I've tracked down is colons in date strings. If you supply a string with colons the signature gets generated (with colons being encoded as `%3A`) but when the request is made the colons are left intact as colon characters.

According to twitter's oauth [documentation](https://dev.twitter.com/oauth/overview/percent-encoding-parameters) colons do not need to be encoded so perhaps the oauth library is incorrect?

@sferik Any intuition as to who which library is "responsible" for this problem?

**Comment by thomas-holmes** [comment]
A simple interim fix seems be to just roll the http dependency back to `0.7.x`

**Comment by sferik** [comment]
> @sferik Any intuition as to who which library is "responsible" for this problem?

I would open an [issue with the `http` gem](https://github.com/httprb/http/issues/new), referencing the [Twitter percent encoding documentation](https://dev.twitter.com/oauth/overview/percent-encoding-parameters) and the [URI RFC](http://tools.ietf.org/html/rfc3986#section-2.1).

It looks like `URI.escape` does the right thing but `URI.encode_www_form`, which is used on line 129 of `lib/http/client.rb` in the `http` gem, does the wrong thing with respect to escaping colons. Reverting back to `http` version 0.7 isn’t a good long-term solution. If this is a bug in the `http` gem, let’s make sure it gets fixed there.

**Comment by ixti** [comment]
Issue with escaping colon fixed in `http` gem (`0.8.13`, `0.9.1`, master branch)

**Comment by stve** [comment]
@thomas-holmes Based on the discussion in #671, we've decided it makes more sense to point to Twitter's SDK. Undoubtedly your work here reflects a considerable amount of effort and while we aren't going to merge this in right now, I'd be remiss if I didn't thank you for opening this pull request and bringing this very close to the finish line. Thanks for contributing! :100:

**Pull request closed without merging**